### ISSUE TITLE
legacy_compound_view::tri_comparator() ignores the prefix when comparing views of different size

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#6814]
While looking into #6797 I looked for similar instances and realized that
https://github.com/scylladb/scylla/blob/4c221855a14dd3107efa14bd17b53433595dd1f7/compound_compat.hh#L151-L154

compares only the bytes_view sizes when they are not equal without memcmp'ing the prefix of the minimum size.

Looks like it could just use compare_unsigned for each component, given #6797 is fixed.

**Comment by bhalevy** [comment]
I'm testing the following change:
```diff
diff --git a/bytes.hh b/bytes.hh
index 9d1a0689b9..2999abde01 100644
--- a/bytes.hh
+++ b/bytes.hh
@@ -84,9 +84,12 @@ struct appending_hash<bytes_view> {
 };
 
 inline int32_t compare_unsigned(bytes_view v1, bytes_view v2) {
-    auto n = memcmp(v1.begin(), v2.begin(), std::min(v1.size(), v2.size()));
+  auto size = std::min(v1.size(), v2.size());
+  if (size) {
+    auto n = memcmp(v1.begin(), v2.begin(), size);
     if (n) {
         return n;
     }
+  }
     return (int32_t) (v1.size() - v2.size());
 }
diff --git a/compound_compat.hh b/compound_compat.hh
index e8e8b03f1d..7783b3823c 100644
--- a/compound_compat.hh
+++ b/compound_compat.hh
@@ -148,10 +148,7 @@ class legacy_compound_view {
                 _type.begin(k1), _type.end(k1),
                 _type.begin(k2), _type.end(k2),
                 [] (const bytes_view& c1, const bytes_view& c2) -> int {
-                    if (c1.size() != c2.size()) {
-                        return c1.size() < c2.size() ? -1 : 1;
-                    }
-                    return memcmp(c1.begin(), c2.begin(), c1.size());
+                    return compare_unsigned(c1.begin(), c2.begin());
                 });
         }
     };
```

**Comment by bhalevy** [comment]
Cc @dekimir @nyh

**Comment by dekimir** [comment]
Nice catch, @bhalevy.  This change would also fix #6797 for everyone, right?

**Comment by bhalevy** [comment]
Sooo, the first part of it would fix #6335 
But, sadly, it looks I might have been wrong about this issue and the comparison of legacy views.

```
test/boost/compound_test.cc(257): Entering test case "test_legacy_ordering_of_composites"
test/boost/compound_test.cc(272): fatal error: in "test_legacy_ordering_of_composites": critical check cmp(make("AA", "B"), make("B", "B")) > 0 has failed
```

The documentation in the code is unclear about how the legacy views should be compared.
Maybe there are a couple of obscure hints about the comparison:
```
    static_assert(!CompoundType::is_prefixable, "Legacy view not defined for prefixes");

    // A trichotomic comparator defined on @CompoundType representations which
    // orders them according to lexicographical ordering of their corresponding
    // legacy representations.
    //
    //   tri_comparator(t)(k1, k2)
    //
    // ...is equivalent to:
    //
    //   compare_unsigned(to_legacy(t, k1), to_legacy(t, k2))
    //
    // ...but more efficient.
    //
```

**Comment by bhalevy** [comment]
Closing this issue since the problem I thought I spotted is apparently not a problem after all.
